planner/cascades/transformation_rules.go

```diff
@@ -589,3 +589,86 @@ func (r *PushSelDownJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.Grou
 	newJoinExpr.SetChildren(leftGroup, rightGroup)
 	return []*memo.GroupExpr{newJoinExpr}, true, false, nil
 }
+
+// MergeAdjacentProjection merge the adjacent projection.
+type MergeAdjacentProjection struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentProjection creates a new Transformation MergeAdjacentProjection.
+// The pattern of this rule is `Projection -> Projection`.
+func NewRuleMergeAdjacentProjection() Transformation {
+	rule := &MergeAdjacentProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandProjection,
+		memo.EngineAll,
+		memo.NewPattern(memo.OperandProjection, memo.EngineAll),
+	)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `proj->proj->x` to `proj->x`
+// or just keep the selection unchanged.
+func (r *MergeAdjacentProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	proj := old.GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	child := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	if plannercore.ExprsHasSideEffects(child.Exprs) {
+		return nil, false, false, nil
+	}
+
+	replace := make(map[string]*expression.Column)
+	for i, col := range child.Schema().Columns {
+		if col1, ok := child.Exprs[i].(*expression.Column); ok {
+			replace[string(col.HashCode(nil))] = col1
+		} else {
+			return nil, false, false, nil
+		}
+	}
+
+	for i, expr := range proj.Exprs {
+		resolveExprAndReplace(expr, replace)
+		proj.Exprs[i] = replaceColumnOfExpr(proj.Exprs[i], child)
+	}
+
+	childGroup := old.Children[0].GetExpr().Children[0]
+	old.GetExpr().SetChildren(childGroup)
+	return nil, false, false, nil
+}
+
+func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
+	dst := replace[string(origin.HashCode(nil))]
+	if dst != nil {
+		retType, inOperand := origin.RetType, origin.InOperand
+		*origin = *dst
+		origin.RetType, origin.InOperand = retType, inOperand
+	}
+}
+
+func resolveExprAndReplace(origin expression.Expression, replace map[string]*expression.Column) {
```

Comment:
Oh...I guess the codes in `rule_eliminate_project.go` would be deprecated when the new cascades planner is finished. So I did the copy-and-paste.

So it would be better to export `resolveExprAndReplace` ?

Reply:
Yes, those codes may be deprecated when the cascades planner is powerful enough. But to reduce the cost of maintaining two optimizers in the recent releases, it is better to export the old function.

Reply:
This function is some historical codes based on some not very serious assumption. Will be removed once other historical codes are fixed.
If you're interested in what's wrong with them, we can have a detailed explaination in slack channel~

Reply:
I see, thanks~~